File: DeepSpace-Java/src/deepspace/NumericDamage.java
package deepspace;

import java.util.ArrayList;

public class NumericDamage extends Damage {
    private int nWeapons;

//
// ──────────────────────────────────────────────────────────── CONSTRUCTORES ─────
//

    @Override
    NumericDamage (int w, int s) {
        super(s);
        nWeapons = w;
    }

    @Override
    NumericDamage (NumericDamage dano) {
        super( dano.getNShields() );
        nWeapons = dano.getNWeapons();
    }

//
// ─────────────────────────────────────────────────────────────── UTILIDADES ─────
//

    @Override
    NumericDamage copy (NumericDamage dano) {
        return new NumericDamage(dano);
    }

    @Override
    public void discardWeapon () {
        if (nWeapons > 0)
            nWeapons--;
    }

    @Override
    public Boolean hasNoEffect () {
        return super.hasNoEffect() && nWeapons == 0;
    }

    @Override
    public NumericDamage adjust (ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int min_dano = Math.min(nWeapons, w.size());

        return new NumericDamage(min_dano, super.adjust_shields(s));
    }

//
// ─────────────────────────────────────────────────────────────── INTERFACES ─────
//

    @Override
    public int getNWeapons () {
        return nWeapons;
    }

    @Override
    public String toString() {
        return super.toString() + "\n\t-> Daño a armas: " + nWeapons;
    }

    @Override
    DamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }

    @Override
    public void discardWeapon(Weapon w) {
        if(nWeapons>0)
            nWeapons--;
    }
}

File: DeepSpace-Ruby/lib/NumericDamage.rb
# encoding:utf-8

require_relative "Damage"
require_relative "NumericDamageToUI"

module Deepspace
    class NumericDamage < Damage
        
        def initialize(w,s)
            super(s)
            @nWeapons = w
        end

        def copy
            NumericDamage.new(@nWeapons, @nShields)
        end

        attr_reader :nWeapons

        def adjust(w,s)
            NumericDamage.new([@nWeapons,w.length].min, super(s))
        end

        def discardWeapon(w)
            if @nWeapons > 0
                @nWeapons -= 1
            end
        end

        def hasNoEffect
            if @nWeapons == 0
                super
            else
                false
        end

        def getUIversion
            NumericDamageToUI.new(self)
        end
        
        def to_s
            getUIversion.to_s
        end
        
        


    end

end
